Integrate jpo-utils Kafka Setup into CV Manager via Submodule

dmccoystephenson opened the pull request:
# PR Details
## Description
### Problem
Some topics created by the `jpo-utils` repository are not currently created by the CV Manager. Relying on separate Kafka initialization methods increases overhead and the potential for bugs.

### Solution
The `jpo-utils` repository has been added as a Git submodule. Its `docker-compose-kafka.yml` file is now referenced by `docker-compose-intersection.yml`. The `kafka` and `kafka_init` services have been removed from the CV Manager setup, which now relies on the `kafka` and `kafka-setup` services provided by `jpo-utils`.

## How Has This Been Tested?
The CV Manager was spun up locally w/ Docker and Kafka was verified to start up correctly and create topics.

## Types of changes
- [ ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)
- [x] Local deployment change

## Checklist:
- [ ] My changes require new environment variables:
  - [ ] I have updated the docker-compose, K8s YAML, and all dependent deployment configuration files.
- [ ] My changes require updates to the documentation:
  - [ ] I have updated the documentation accordingly.
- [ ] My changes require updates and/or additions to the unit tests:
  - [ ] I have modified/added tests to cover my changes.
- [ ] All existing tests pass.

dmccoystephenson committed:
Add jpo-utils as a new submodule

This commit introduces the jpo-utils repository as a git submodule. It is located at the path `jpo-utils` and links to the remote repository `https://github.com/usdot-jpo-ode/jpo-utils`.

dmccoystephenson committed:
Refactor docker-compose-intersection.yml to use shared Kafka configuration

Removed local kafka and kafka_init service definitions and replaced them by referencing kafka and kafka-setup from the shared docker-compose-kafka.yml. Updated service dependencies and profiles to align with the modular setup, improving reusability and maintainability.

dmccoystephenson commented on sample.env:
```diff
@@ -4,8 +4,8 @@
 # The currently available profile groups are listed below. 
-# basic, webapp, intersection, intersection_no_api, conflictmonitor, addons, obu_ota
-COMPOSE_PROFILES=basic,webapp,intersection
+# basic, webapp, kafka_full, intersection, intersection_no_api, conflictmonitor, addons, obu_ota
+COMPOSE_PROFILES=basic,webapp,kafka_full,intersection
```
Note: Rather than assigning the intersection profile to the jpo-utils Kafka service, I've included kafka_full in the default Compose profiles by default.

dmccoystephenson reviewed (commented):
Self-review

dmccoystephenson committed:
Removed `kafka_init.sh` script

Michael7371 reviewed (commented):
Have you considered removing the CV Manager MongoDB related services in favor of the JPO utils services?

dmccoystephenson commented:
> Have you considered removing the CV Manager MongoDB related services in favor of the JPO utils services?

Yes! I have opened [a separate draft PR](https://github.com/CDOT-CV/jpo-cvmanager/pull/182) for the Mongo side of things. I've still got a bit of troubleshooting to do there.

Michael7371 commented:
> > Have you considered removing the CV Manager MongoDB related services in favor of the JPO utils services?
> 
> Yes! I have opened [a separate draft PR](https://github.com/CDOT-CV/jpo-cvmanager/pull/182) for the Mongo side of things. I've still got a bit of troubleshooting to do there.

Ahh okay, that makes sense!

jacob6838 reviewed (changes requested):
Tested locally - worked great! 
Can you add some documentation to the README about the new jpo-utils submodule and how to use it (the included .env)?

dmccoystephenson committed:
Update README.md to clarify Kafka setup and initialization steps

Renamed `kafka_init` to `kafka-setup` and added notes linking Kafka services to the jpo-utils repository. Expanded the submodule initialization and environment setup instructions for clarity, and updated the steps for running the CV Manager to include changes to Kafka-related configurations.

dmccoystephenson commented:
> Tested locally - worked great! Can you add some documentation to the README about the new jpo-utils submodule and how to use it (the included .env)?

Added!

Michael7371 commented on README.md:
````diff
@@ -151,11 +153,16 @@ Some simple sample data is injected into the MongoDB instance when created. If m
+    ```sh
+    git submodule update --init --recursive
+    ```
+3. Create a copy of the sample.env named ".env" and refer to the Environmental variables section below for more information on each variable.
````
The creation of the `.env` file in the `/jpo-utils` folder is no longer required if we opt to use the CDOT fork of `jpo-utils`. We would want to eventually swap over to the USDOT one during the release but figured I'd let you know.

dmccoystephenson replied:
@payneBrandon Should we point to CDOT jpo-utils or USDOT jpo-utils in this PR?

payneBrandon replied:
Is that one getting merged into usdot with this release?

dmccoystephenson replied:
Yes it should be

dmccoystephenson replied:
Addressed here: https://github.com/CDOT-CV/jpo-cvmanager/pull/194

Michael7371 reviewed (approved):
Looks good, just a comment on potentially using the CDOT fork for now until the changes get merged into USDOT.

drewjj reviewed (approved):
Looks like a small merge conflict with develop is present, but it runs well.

dmccoystephenson committed:
Merge remote-tracking branch 'origin/develop' into docker/reference-kafka-in-jpo-utils

# Conflicts:
#	.gitmodules

dmccoystephenson committed:
Update service dependency from kafka_init to kafka-setup

Renamed the service dependency in docker-compose to align with the updated naming convention.

dmccoystephenson committed:
Changed the jpo-utils submodule URL to point to the `CDOT-CV` repository instead of the `usdot-jpo-ode` repository

dmccoystephenson committed:
Merge pull request #194 from CDOT-CV/docker/reference-kafka-in-jpo-utils-switch-to-cdot-fork

Update jpo-utils submodule to use CDOT fork

Michael7371 commented:
@dmccoystephenson I just realized that the CDOT jpo-utils reference that was pulled in is fairly out of date and doesn't include recent updates. Can you update the commit reference to be: [0c936aa154218fc8b191a0832e2459e2205fa679](https://github.com/CDOT-CV/jpo-utils/commit/0c936aa154218fc8b191a0832e2459e2205fa679) (the latest commit on the develop branch)

dmccoystephenson committed:
Switched jpo-utils submodule to latest commit on `cdot-release_2025-q2` branch

dmccoystephenson commented:
> @dmccoystephenson I just realized that the CDOT jpo-utils reference that was pulled in is fairly out of date and doesn't include recent updates. Can you update the commit reference to be: [0c936aa154218fc8b191a0832e2459e2205fa679](https://github.com/CDOT-CV/jpo-utils/commit/0c936aa154218fc8b191a0832e2459e2205fa679) (the latest commit on the develop branch)

Nice catch! I switched jpo-utils submodule to latest commit on `cdot-release_2025-q2` branch since it was a bit more updated than `develop`.

Michael7371 commented on README.md:
```diff
@@ -151,11 +153,16 @@ Some simple sample data is injected into the MongoDB instance when created. If m
     1.  Make sure at least the DOCKER_HOST_IP, KEYCLOAK_ADMIN_PASSWORD, KEYCLOAK_API_CLIENT_SECRET_KEY, and MAPBOX_TOKEN are set for this.
     2.  Some of these variables, delineated by sections, pertain to the [jpo-conflictmonitor](https://github.com/usdot-jpo-ode/jpo-conflictmonitor), [jpo-geojsonconverter](https://github.com/usdot-jpo-ode/jpo-geojsonconverter), and [jpo-ode](https://github.com/usdot-jpo-ode/jpo-ode). Please see the documentation provided for these projects when setting these variables.
-3.  The CV Manager has four components that need to be containerized and deployed: the API, the PostgreSQL database, Keycloak, and the webapp.
+4. Create a copy of the jpo-utils/sample.env file named "jpo-utils/.env" and set the DOCKER_HOST_IP environment variable. 
```
This shouldn't be required anymore with the latest jpo-utils version.

dmccoystephenson replied:
Removed!

Michael7371 reviewed (approved):
Just a documentation comment, otherwise it LGTM.

dmccoystephenson committed:
Removed unnecessary instruction in README
